dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/ingest/UningestService.java

```diff
@@ -0,0 +1,189 @@
+    private DatasetVersionRepository datasetVersionRepository;
+    private DatasetVersionServiceBean datasetVersionService;
+
+    // -------------------- CONSTRUCTORS --------------------
+
+    public UningestService() {
+        this.dataAccess = DataAccess.dataAccess();
+    }
+
+    @Inject
+    public UningestService(DataTableRepository dataTableRepository, IngestReportRepository ingestReportRepository,
+                           DataFileTagRepository dataFileTagRepository, DataFileRepository dataFileRepository,
+                           FileMetadataRepository fileMetadataRepository, MapLayerMetadataServiceBean mapLayerMetadataService,
+                           DatasetVersionRepository datasetVersionRepository, DatasetVersionServiceBean datasetVersionService) {
+        this();
+        this.mapLayerMetadataService = mapLayerMetadataService;
+        this.datasetVersionRepository = datasetVersionRepository;
+        this.dataTableRepository = dataTableRepository;
+        this.ingestReportRepository = ingestReportRepository;
+        this.dataFileTagRepository = dataFileTagRepository;
+        this.dataFileRepository = dataFileRepository;
+        this.fileMetadataRepository = fileMetadataRepository;
+        this.datasetVersionService = datasetVersionService;
+    }
+
+    // -------------------- LOGIC --------------------
+
+    public void uningest(List<DataFile> files, AuthenticatedUser user) {
+        if (files == null || files.isEmpty() || user == null) {
+            return;
+        }
+        DataFile anyMergedFile = null;
+        for (DataFile file : files) {
+            anyMergedFile = uningestFile(file, user);
+        }
+        // As all files are from the same version it's sufficient to use
+        // any of the files to recalculate UNF
+        recalculateUNFForCurrentVersion(anyMergedFile);
+    }
+
+    // -------------------- PRIVATE --------------------
+
+    private DataFile uningestFile(DataFile file, AuthenticatedUser user) {
+        // File to uningest cannot be already present in any published version
+        if (file.getFileMetadatas().size() > 1 || !file.getFileMetadata().getDatasetVersion().isDraft()) {
+            throw new IllegalArgumentException("Only fresh files from draft can be uningested!");
+        }
+        // Variable file is updated in order to avoid OptimisticLockException
+        file = file.isTabularData()
+                ? uningestTabularFile(file)
+                : uningestFileWithoutDataTable(file);
+        return removeWorldMapData(file, user);
+    }
+
+    private DataFile uningestTabularFile(DataFile file) {
+        long originalSize;
+        StorageIO<DataFile> storage;
+        try {
+            storage = dataAccess.getStorageIO(file);
+            storage.open();
+            originalSize = storage.getAuxObjectSize(StorageIOConstants.SAVED_ORIGINAL_FILENAME_EXTENSION);
+            storage.revertBackupAsAux(StorageIOConstants.SAVED_ORIGINAL_FILENAME_EXTENSION);
```

Comment:
IMO it would be safer to make changes on storage at the end. When operations on db has been sucessfully executed. Do note that we are able to revert changes on database (transaction rollback). But reverting changes in storage is not so easy.
I'm not insisting on this, but still I think it would be beneficial.
What do you think?